### PR TITLE
histogram: Modify getBound to deal properly with infinity

### PR DESCRIPTION
**Pull request by beorn7**
Follow-up PR to https://github.com/prometheus/client_golang/pull/1144 , now that the ±Inf behavior is clarified.

The bucket receiving math.MaxFloat64 observations now has math.MaxFloat64 as upper bound, while the bucket after it (the last possible bucket) has +Inf.

This also adds a test for getBound and moves the getBound code to generic.go (where it should have been in the first place).

### CODE REVIEW

File: model/histogram/generic.go [hunk 1]
```diff
@@ -325,3 +326,186 @@ func compactBuckets[IBC internalBucketCount](buckets []IBC, spans []Span, maxEmp
 
 	return buckets, spans
 }
+
+func getBound(idx, schema int32) float64 {
```
Comment (codesome) [inline model/histogram/generic.go:330]: Where do the numbers 1024 1025 and frac==0.5 come from? I feel it is not trivial to get my head around the math here.

File: model/histogram/generic_test.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+			want:   math.Inf(+1),
+		},
+	}
+
+	for i, s := range scenarios {
+		got := getBound(s.idx, s.schema)
+		if s.want != got {
+			t.Errorf("%d. idx %d, schema %d, want upper bound of %g, got %g", i, s.idx, s.schema, s.want, got)
+		}
+	}
```
Comment (codesome) [inline model/histogram/generic_test.go:109]: We can use `require.Equal` like other tests; that allows writing custom messages as well.

**Review by codesome — APPROVED** [review]
LGTM with a comment and a question

File: model/histogram/generic_test.go (hunk 2 above)
Comment (beorn7) [reply model/histogram/generic_test.go:109]: Right. I have just copied this from the corresponding client_golang code, where we refrain from using higher-level test frameworks. Will fix.

File: model/histogram/generic.go (hunk 1 above)
Comment (beorn7) [reply model/histogram/generic.go:330]: Yeah, it requires a lecture about IEEE 754 floats, essentially. I'll try my best…

**Pull request merged**